5.4.1 SQL AND ETL Challenges

I select SET 2 for this week 5 Assessment

In [1]:
#import libraries
import pandas as pd
import numpy as np

In [2]:
#read crime.csv
data = pd.read_csv('Desktop/crime.csv')

In [4]:
data.head()

cdatetime              address  district        beat  grid  \
0  1/1/2006 0:00   3108 OCCIDENTAL DR         3  3C          1115   
1  1/1/2006 0:00  2082 EXPEDITION WAY         5  5A          1512   
2  1/1/2006 0:00           4 PALEN CT         2  2A           212   
3  1/1/2006 0:00       22 BECKFORD CT         6  6C          1443   
4  1/1/2006 0:00     3421 AUBURN BLVD         2  2A           508   

                      crimedescr  ucr_ncic_code   latitude   longitude  
0  10851(A)VC TAKE VEH W/O OWNER           2404  38.550420 -121.391416  
1     459 PC  BURGLARY RESIDENCE           2204  38.473501 -121.490186  
2  10851(A)VC TAKE VEH W/O OWNER           2404  38.657846 -121.462101  
3   476 PC PASS FICTICIOUS CHECK           2501  38.506774 -121.426951  
4   459 PC  BURGLARY-UNSPECIFIED           2299  38.637448 -121.384613

Set 2 FEATURE ENGINEERING

Using data found in crime.csv, perform the following tasks:

Task 1: Give the average number of characters of the entries of the crimedescr variable

In [5]:
# Calculate the total number of characters
total_characters = sum(len(entry) for entry in data['crimedescr'])

# Calculate the average number of characters
average_characters = total_characters / len(data['crimedescr'])

average_characters

24.883043248945146

Ave = 24.88

Task 2: Scale the latitude and longitude using the min-max normalization method

Using formula for min-max normalization
x' = (x - min(x)) / (max(x) - min(x))

In [6]:
# Define a function for min-max normalization
def min_max_normalize(column):
    min_val = column.min()
    max_val = column.max()
    normalized_column = (column - min_val) / (max_val - min_val)
    return normalized_column

In [7]:
# Apply min-max normalization to the latitude and longitude columns
data['latitude_normalized'] = min_max_normalize(data['latitude'])
data['longitude_normalized'] = min_max_normalize(data['longitude'])

# Display the DataFrame with the normalized columns
data[['latitude_normalized', 'longitude_normalized']]

latitude_normalized  longitude_normalized
0                0.457386              0.862651
1                0.144437              0.344430
2                0.894448              0.491785
3                0.279809              0.676208
4                0.811458              0.898346
...                   ...                   ...
7579             0.568910              0.447929
7580             0.405188              0.331809
7581             0.545307              0.466018
7582             0.362859              0.443794
7583             0.169366              0.142700

[7584 rows x 2 columns]

The output values are in 0.0 - 1.0 range

Task 3: Identify the number of crimes that occured in the weekends

In [8]:
# Convert the 'cdatetime' column to datetime format
data['cdatetime'] = pd.to_datetime(data['cdatetime'])

# Extract the day of the week (0 = Monday, 6 = Sunday)
data['day_of_week'] = data['cdatetime'].dt.dayofweek

# Define a function to check if the day is a weekend (Saturday or Sunday)
def is_weekend(day):
    return day in [5, 6]  # Saturday and Sunday

# Use the function to filter crimes that occurred on weekends
weekend_crimes = data[data['day_of_week'].apply(is_weekend)]

# Count the number of crimes that occurred on weekends
num_weekend_crimes = len(weekend_crimes)

num_weekend_crimes

1945

Number of crimes in weekends = 1945